### PR TITLE
Add support for extra claims

### LINKED ISSUE #496: Add support for extra_scopes
There is a possibility to support extra scopes. This should be easily possible.

Proposal:
* Add `extra_scope_dict` kwarg to `oic.oic.Provider` and use them in `ProviderConfiguration`
* Remove extraneous `openid` scope
* Add `extra_scope_dict` kwarg to `oic.oic.scope2claims`
* Pass this kwarg to `scope2claims` in `oic.oic.Provider._collect_user_info()`

### PR DESCRIPTION
**Pull request by tpazderka**
Close #496

- [x] Any changes relevant to users are recorded in the `CHANGELOG.md`.
- [x] The documentation has been updated, if necessary.
---

### CODE REVIEW

File: doc/examples/op.rst [hunk 1]
```diff
@@ -143,3 +143,16 @@ formatted html. ``template_name`` can be one of following:
     and ``inputs`` a dictionary with passed attributes.
 
   * ``verify_logout`` - used for verification of logout at Provider.
+
+
+Extension of supported claims and scopes
+----------------------------------------
+To extend the supported and recognized claims and scopes, it is possible to pass a keyword argumets ``extra_claims`` and
```
[the remark this answers is not in the document]
Comment (tpazderka) [reply doc/examples/op.rst]: There are two of them. `extra_claims` and `extra_scope_dict` (wrapped on next line...)

[the remark this answers is not in the document]
Comment (schlenk) [reply doc/examples/op.rst]: Well, the spelling error, not the plural ;-)

[the remark this answers is not in the document]
Comment (tpazderka) [reply doc/examples/op.rst]: :) Fixed.

File: src/oic/oic/__init__.py [hunk 2]
```diff
@@ -1818,11 +1818,15 @@ def make_id_token(self, session, loa="2", issuer="",
         return idt
 
 
-def scope2claims(scopes):
+def scope2claims(scopes, extra_scope_dict=None):
```
Comment (schlenk) [inline src/oic/oic/__init__.py]: Looking at the method, it is surprising it doesn't simply return a set() and instead creates this dict with all values empty. Not sure if any caller uses the values in a meaningful way...

File: src/oic/oic/provider.py [hunk 3]
```diff
@@ -1681,17 +1682,24 @@ def provider_features(self, pcr_class=ProviderConfigurationResponse):
 
         _provider_info = pcr_class(**CAPABILITIES)
 
+        # Parse scopes
+        _scopes = list(SCOPE2CLAIMS.keys())
+        if self.extra_scope_dict is not None:
+            _scopes.extend(self.extra_scope_dict.keys())
+        # Remove duplicates if any
+        _provider_info["scopes_supported"] = list(set(_scopes))
```
Comment (schlenk) [inline src/oic/oic/provider.py]: You can actually write this shorter and without a copy...

_provider_info[...] = list(set(itertools.chain(SCOPE2CLAIMS.keys(), self.extra_scope_dict.keys() if self.extra_scope_dict else []))

but i think your version is more readable ;-)

File: doc/examples/op.rst [hunk 4]
```diff
@@ -143,3 +143,16 @@ formatted html. ``template_name`` can be one of following:
     and ``inputs`` a dictionary with passed attributes.
 
   * ``verify_logout`` - used for verification of logout at Provider.
+
+
+Extension of supported claims and scopes
+----------------------------------------
+To extend the supported and recognized claims and scopes, it is possible to pass a keyword arguments ``extra_claims``
```
[the remark this answers is not in the document]
Comment (tpazderka) [reply doc/examples/op.rst]: Fixed

File: src/oic/oic/provider.py (hunk 3 above)
Comment (tpazderka) [reply src/oic/oic/provider.py]: I will stick to mine, I guess readability is more important... :)

File: src/oic/oic/__init__.py (hunk 2 above)
Comment (tpazderka) [reply src/oic/oic/__init__.py]: The method is used in only one place so I guess the answer is no :)

Looking at the code it is a legacy thing...

If you want, we can make a ticket for refactoring of this thing...

**Comment by schlenk** [comment]
Feel free to merge.

**Pull request merged**